cpp/src/core/bitset.cu

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright (c) 2023-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuvs/core/bitset.hpp>
+#include <raft/core/detail/mdspan_util.cuh>  // native_popc
+#include <raft/core/device_container_policy.hpp>
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/resource/thrust_policy.hpp>
+#include <raft/core/resources.hpp>
+#include <raft/linalg/map.cuh>
+#include <raft/linalg/reduce.cuh>
+#include <raft/util/device_atomics.cuh>
+#include <thrust/for_each.h>
+
+namespace cuvs::core {
+
+template <typename bitset_t, typename index_t>
+_RAFT_HOST_DEVICE bitset_view<bitset_t, index_t>::bitset_view(bitset_t* bitset_ptr,
+                                                              index_t bitset_len)
+  : bitset_ptr_{bitset_ptr}, bitset_len_{bitset_len}
+{
+}
+
+template <typename bitset_t, typename index_t>
+_RAFT_HOST_DEVICE bitset_view<bitset_t, index_t>::bitset_view(
+  raft::device_vector_view<bitset_t, index_t> bitset_span, index_t bitset_len)
+  : bitset_ptr_{bitset_span.data_handle()}, bitset_len_{bitset_len}
+{
+}
+
+template <typename bitset_t, typename index_t>
+_RAFT_DEVICE void bitset_view<bitset_t, index_t>::set(const index_t sample_index,
+                                                      bool set_value) const
+{
+  const index_t bit_element = sample_index / bitset_element_size;
+  const index_t bit_index   = sample_index % bitset_element_size;
+  const bitset_t bitmask    = bitset_t{1} << bit_index;
+  if (set_value) {
+    atomicOr(bitset_ptr_ + bit_element, bitmask);
+  } else {
+    const bitset_t bitmask2 = ~bitmask;
+    atomicAnd(bitset_ptr_ + bit_element, bitmask2);
+  }
+}
+template <typename bitset_t, typename index_t>
+_RAFT_HOST_DEVICE index_t bitset_view<bitset_t, index_t>::n_elements() const
+{
+  return raft::ceildiv(bitset_len_, bitset_element_size);
+}
+template <typename bitset_t, typename index_t>
+raft::device_vector_view<bitset_t, index_t> bitset_view<bitset_t, index_t>::to_mdspan()
+{
+  return raft::make_device_vector_view<bitset_t, index_t>(bitset_ptr_, n_elements());
+}
+template <typename bitset_t, typename index_t>
+raft::device_vector_view<const bitset_t, index_t> bitset_view<bitset_t, index_t>::to_mdspan() const
+{
+  return raft::make_device_vector_view<const bitset_t, index_t>(bitset_ptr_, n_elements());
+}
+
+template <typename bitset_t, typename index_t>
```

Comment:
I don't mind compiling `bitset` into cuvs, but i still think much of the implementation should live in headers in raft. I think what we can do is use the `using` clause in order to expose a `cuvs::bitset` while still having the bitset code officially live in raft. This just makes things more flexible for users of raft.